src/V3Timing.cpp

```diff
@@ -0,0 +1,1190 @@
+private:
+    // NODE STATE
+    //  AstVar::user1()      -> AstNode*.  Event to trigger on var's posedge
+    //  AstVar::user2()      -> AstNode*.  Event to trigger on var's negedge
+    //  AstVar::user3()      -> AstNode*.  Event to trigger on var's anyedge
+    // VNUser1InUse    m_inuser1;      (Allocated for use in TimingTransformVisitor)
+    // VNUser2InUse    m_inuser2;      (Allocated for use in TimingTransformVisitor)
+    // VNUser2InUse    m_inuser3;      (Allocated for use in TimingTransformVisitor)
+
+    // STATE
+    using VarScopeSet = std::set<AstVarScope*>;
+    VarScopeSet m_waitVars;  // Set of vars in wait expression
+
+    bool m_inEventControlSens = false;  // Are we under a event control sens list?
+    bool m_inWait = false;  // Are we under a wait statement?
+    AstSenItem* m_senItemp = nullptr;  // The senitem we're under
+
+    // METHODS
+    VL_DEBUG_FUNC;  // Declare debug()
+
+    // VISITORS
+    virtual void visit(AstEventControl* nodep) override {
+        VL_RESTORER(m_inEventControlSens);
+        m_inEventControlSens = true;
+        iterateAndNextNull(nodep->sensesp());
+        m_inEventControlSens = false;
+        iterateAndNextNull(nodep->stmtsp());
+    }
+    virtual void visit(AstWait* nodep) override {
+        VL_RESTORER(m_inWait);
+        m_inWait = true;
+        iterateAndNextNull(nodep->condp());
+        if (m_waitVars.empty()) {  // There are no vars to wait on
+            if (nodep->bodysp())
+                nodep->replaceWith(nodep->bodysp()->unlinkFrBackWithNext());
+            else
+                nodep->unlinkFrBack();
+        } else {
+            auto fl = nodep->fileline();
+            AstNode* senitemsp = nullptr;
+            // Wait on anyedge events related to the vars in the wait statement
+            for (auto* const vscp : m_waitVars) {
+                AstVarScope* const eventp = vscp->varp()->isEventValue()
+                                                ? vscp
+                                                : getCreateEdgeEvent(vscp, VEdgeType::ET_ANYEDGE);
+                senitemsp = AstNode::addNext(
+                    senitemsp, new AstSenItem{fl, VEdgeType::ET_ANYEDGE,
+                                              new AstVarRef{fl, eventp, VAccess::READ}});
+            }
+            auto* const condp = nodep->condp()->unlinkFrBack();
+            auto* const eventControlp
+                = new AstEventControl{fl, new AstSenTree{fl, VN_AS(senitemsp, SenItem)}, nullptr};
+            // Put the event control in a while loop with the wait statement as condition
+            auto* const whilep = new AstWhile{fl, new AstLogNot{fl, condp}, eventControlp};
+            if (nodep->bodysp()) whilep->addNext(nodep->bodysp()->unlinkFrBackWithNext());
+            nodep->replaceWith(whilep);
+            m_waitVars.clear();
+        }
+        VL_DO_DANGLING(nodep->deleteTree(), nodep);
```

Comment:
```suggestion
        VL_DO_DANGLING(pushDeletep(nodep), nodep);
```

Makes debug easier as node* won't get reused in this visitor.  Generally this is preferred for rare/complicated removals.  For something super-hot not good to do this (e.g. V3Const).  Look for others.